### PR TITLE
RSDK-8819: Finish FTDC

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

File: ftdc/ftdc.go [hunk 1]
```diff
@@ -125,21 +125,33 @@ type FTDC struct {
...
 }
 
-// New creates a new *FTDC.
-func New(logger logging.Logger) *FTDC {
-	return NewWithWriter(nil, logger)
+// New creates a new *FTDC. This FTDC object will write FTDC formatted files into the input
+// `ftdcDirectory`.
+func New(ftdcDirectory string, logger logging.Logger) *FTDC {
+	ret := newFTDC(logger)
+	ret.maxFileSizeBytes = 1_000_000
```
Comment (dgottlieb) [inline ftdc/ftdc.go:132]: Made the settings a more formal division. One either passes in a directory and all these variables about managing files are initialized. Or one passes in a writer and none of those variables matter.

File: ftdc/ftdc.go [hunk 2]
```diff
@@ -536,7 +553,7 @@ func (ftdc *FTDC) checkAndDeleteOldFiles() error {
 // deletion testing. Filename generation uses padding such that we can rely on there before 2/4
 // digits for every numeric value.
 //
-//nolint
+// nolint
```
Comment (dgottlieb) [inline ftdc/ftdc.go:556]: It joys me that our linter lints our linter

File: ftdc/ftdc_test.go [hunk 3]
```diff
@@ -294,18 +294,16 @@ func TestStatsWriterContinuesOnSchemaError(t *testing.T) {
...
+	// Isolate all of the files we're going to create to a single, fresh directory.
 	ftdcFileDir, err := os.MkdirTemp("./", "countingBytesTest")
 	test.That(t, err, test.ShouldBeNil)
 	defer os.RemoveAll(ftdcFileDir)
 
-	// Isolate all of the files we're going to create to a single, fresh directory.
-	ftdc.ftdcDir = ftdcFileDir
+	// We must not use `NewWithWriter`. Forcing a writer for FTDC data is not compatible with FTDC
+	// file rotation.
+	ftdc := New(ftdcFileDir, logger.Sublogger("ftdc"))
```
Comment (dgottlieb) [inline ftdc/ftdc_test.go:304]: The "refactor" of `New` vs `NewWithWriter` resulted in changing the order here. First create the directory, then pass it to the constructor.

File: robot/impl/local_robot.go [hunk 4]
```diff
@@ -416,7 +416,8 @@ func newWithResources(
 
 	var ftdcWorker *ftdc.FTDC
 	if rOpts.enableFTDC {
-		ftdcWorker = ftdc.New(logger.Sublogger("ftdc"))
+		// CloudID is also known as the robot part id.
+		ftdcWorker = ftdc.New(ftdc.DefaultDirectory(config.ViamDotDir, cfg.Cloud.ID), logger.Sublogger("ftdc"))
```
Comment (dgottlieb) [inline robot/impl/local_robot.go]: Technically has a bug -- cfg.Cloud.ID is nil for local configs. Fixed in last commit.

File: ftdc/cmd/parser.go [hunk 5]
```diff
@@ -36,6 +39,31 @@ type gnuplotWriter struct {
 	options graphOptions
 }
 
+type KVPair[K, V any] struct {
```
Comment (dgottlieb) [inline ftdc/cmd/parser.go:42]: I expect you'll enjoy this @benjirewis

File: ftdc/cmd/parser.go [hunk 6]
```diff
@@ -165,25 +194,20 @@ func (gpw *gnuplotWriter) CompileAndClose() string {
...
-
 		nolintPrintln("Error opening file. File:", os.Args[1], "Err:", err)
-
 		nolintPrintln("Expected an FTDC filename. E.g: go run parser.go <path-to>/viam-server.ftdc")
 		return
 	}
 
-	data, err := ftdc.Parse(ftdcFile)
+	logger := logging.NewDebugLogger("parser")
+	logger.SetLevel(logging.WARN)
```
Comment (dgottlieb) [inline ftdc/cmd/parser.go:209]: Replaced with a simple `NewLogger` and removed the `SetLevel` -- using the INFO default.

File: robot/impl/local_robot.go [hunk 7]
```diff
@@ -416,8 +416,12 @@ func newWithResources(
 
 	var ftdcWorker *ftdc.FTDC
 	if rOpts.enableFTDC {
+		partID := "local-config"
+		if cfg.Cloud != nil {
```
Comment (dgottlieb) [inline robot/impl/local_robot.go:420]: bug fix

File: web/server/entrypoint.go [hunk 8]
```diff
@@ -48,7 +48,7 @@ type Arguments struct {
 	OutputTelemetry            bool   `flag:"output-telemetry,usage=print out telemetry data (metrics and spans)"`
 	DisableMulticastDNS        bool   `flag:"disable-mdns,usage=disable server discovery through multicast DNS"`
 	DumpResourcesPath          string `flag:"dump-resources,usage=dump all resource registrations as json to the provided file path"`
-	EnableFTDC                 bool   `flag:"ftdc,usage=enable fulltime data capture for diagnostics [beta feature]"`
+	EnableFTDC                 bool   `flag:"ftdc,default=true,usage=enable fulltime data capture for diagnostics"`
```
Comment (dgottlieb) [inline web/server/entrypoint.go:51]: Stole from `WebRTC` above on line 45. Hand tested that `-ftdc=false` turns this off.

File: ftdc/ftdc.go (hunk 2 above)
Comment (dgottlieb) [reply ftdc/ftdc.go]: Or maybe this was gofmt. The linter complained and it's now back to its unindented state.

File: ftdc/cmd/parser.go [hunk 9]
```diff
@@ -36,6 +39,31 @@ type gnuplotWriter struct {
 	options graphOptions
 }
 
+type kvPair[K, V any] struct {
+	Key K
+	Val V
+}
+
+func sorted[K cmp.Ordered, V any](mp map[K]V) []kvPair[K, V] {
```
Comment (benjirewis) [inline ftdc/cmd/parser.go:47]: Hah love it. Normally I've seen
1. Build a slice of the keys
2. `sort.Strings` that slice 
3. Iterate through sorted slice to access map in key order

But this definitely also works.

File: ftdc/cmd/parser.go [hunk 10]
```diff
@@ -165,25 +194,19 @@ func (gpw *gnuplotWriter) CompileAndClose() string {
...
-
 		nolintPrintln("Error opening file. File:", os.Args[1], "Err:", err)
-
 		nolintPrintln("Expected an FTDC filename. E.g: go run parser.go <path-to>/viam-server.ftdc")
 		return
 	}
 
-	data, err := ftdc.Parse(ftdcFile)
+	logger := logging.NewLogger("parser")
+	data, err := ftdc.ParseWithLogger(ftdcFile, logger)
```
Comment (benjirewis) [inline ftdc/cmd/parser.go:209]: Does the parser `main` program here get a lot of info-level-and-above logs now?

File: ftdc/ftdc.go [hunk 11]
```diff
@@ -415,6 +429,15 @@ func (ftdc *FTDC) getWriter() (io.Writer, error) {
 	// It's unclear in what circumstance we'd expect creating a new file to fail. Try 5 times for no
 	// good reason before giving up entirely and shutting down FTDC.
 	for numTries := 0; numTries < 5; numTries++ {
+		// The viam process is expected to be run as root. The FTDC directory must be readable by
+		// "other" users.
+		//
+		//nolint:gosec
+		if err = os.MkdirAll(ftdc.ftdcDir, 0o755); err != nil {
```
Comment (benjirewis) [inline ftdc/ftdc.go:436]: Won't this recreate the FTDC directory on every attempt to create a single new FTDC file?

File: robot/impl/local_robot.go (hunk 7 above)
Comment (benjirewis) [reply robot/impl/local_robot.go:420]: What bug was this? Just looks like starting FTDC within a `~/.viam/[part-id|"local-config"]` as a new feature.

File: robot/web/web.go [hunk 12]
```diff
@@ -188,6 +188,8 @@ type Service interface {
 
 	// Returns the unix socket path the module server listens on.
 	ModuleAddress() string
+
+	Stats() any
```
Comment (benjirewis) [inline robot/web/web.go:192]: Surprised linter does not care about lack of comment here.

**Review by benjirewis — COMMENTED** [review]
Nice! Some questions.

File: robot/impl/local_robot.go (hunk 7 above)
Comment (benjirewis) [reply robot/impl/local_robot.go:420]: Oh nvm I see your other comment.

File: ftdc/ftdc.go (hunk 11 above)
Comment (dgottlieb) [reply ftdc/ftdc.go:436]: MkdirAll has the same behavior as `mkdir -p`. Where it's a no-op + no error if the directory already exists.

FWIW, the case you're describing is exercised in a test.

File: ftdc/cmd/parser.go (hunk 10 above)
Comment (dgottlieb) [reply ftdc/cmd/parser.go:209]: IIRC, there are no info logs. Just debug for item by item details and warn/error for problems.

**Review by benjirewis — APPROVED** [review]
Enable away! 🚀

**Pull request merged**